### DISCUSSION TITLE
Map resource configuration to different instance of the same `op`

### DISCUSSION THREAD

**Discussion by ppasquet** [dagster-io/dagster#19201] [Q&A]
Hi, I am working on my first dagster flow and trying to understand what's the best practice for my use-case.

TL;DR - Trying to create a graph of two identical API job, where only the API endpoint varies. 

I have a configurable resource that acts as an interface to the data provider. 
```
class MyAPI(ConfigurableResource):
    url: str = Field()
    ...
```

I have simplified the execution flow for the sake of clarity, but basically I am HTTP-requesting an endpoint and then processing the results. 

```
@op 
def query_api():
    # this op calls a method from MyAPI

@op
def process_1():
     ...

@op
def process_2():
    ...

@graph
def get_data():
    query_api()
    process_1()
    process_2()
```

Now, my expected workflow is to launch `get_data()` on two different endpoints (different `url` in `MyAPI`). 

```
@graph
def workflow():
    get_data.alias("op1") # for endpoint 1
    get_data.alias("op2") # for endpoint 2
    # Regroup the two datasets ... 
```

In this example, I am only querying two different endpoint but in practice the number of endpoint is unbounded. All "branches" of the workflow are identical, except for URL. 

**Therefore, I'd like to map a different `MyAPI` configuration for each `op` (defined by `get_data()`).**

This is my `Definitions`

```
defs = Definitions(
    jobs=[
        workflow.to_job(
            resource_defs={
                "api_resource": MyAPI()
           }
        )
    ]
)
```

1. Is this the right way of doing things? Basically, whenever `workflow` is triggered, all branchs (endpoint) are ran in parallel, and their results processed altogether once they're all done. 
2. If it is, I imagine I need to specify the mapping in the `Definitions`? Something like
```
defs = Definitions(
    jobs=[
        workflow.to_job(
            resource_defs={
                "api_resource": {
                     "op1": MyAPI("endpoint1"),
                     "op2": MyAPI("endpoint2"),
                 }
           }
        )
    ]
)
```
3. If it's not, how should I structure my workflow?

Thanks!

**Comment by OwenKephart** [comment]
Hi @ppasquet -- unfortunately, the resource key is the unique identifier for a resource in a job (and so there will only be a single instance of resource config for any particular resource key).

One potential recommendation would be to remove the endpoint from your API config and instead just read the alias off of the context object within your op:

```python
@op
def my_op(context, resource: MyAPI):
    MyAPI.do_thing(endpoint=context.op.name)

my_op.alias("endpoint1")
my_op.alias("endpoint2")
```

If your endpoint names are not well-formed, and so would not make good aliases, another option would be a simple op factory:

```python
def get_my_op(endpoint):
    @op(name=<unique_identfier_here>)
    def _op(resource: MyAPI):
        MyAPI.do_thing(endpoint=endpoint)
    return _op

get_my_op("endpoint1")
get_my_op("endpoint2")
```

**Reply by ppasquet** [reply]
Thanks @OwenKephart - this makes sense, although in practice it fails short as `context.op.name` refers to the inner `my_op` , not the overarching graph. I can't propagate the overarching graph alias to the inner op. 

I think I'll explore another route, i.e. sensors.
